File: 211-design-add-search-words-data-structure.ts
class WordDictionary {
  root: TrieNode = new TrieNode()

  addWord(word: string): void {
    let node = this.root
    for (const char of word) {
      const index = CHAR_INDEX[char]
      if (!node.children[index]) {
        node.children[index] = new TrieNode()
      }
      node = node.children[index]!
    }
    node.isTerminator = true
  }

  search(word: string): boolean {
    let cursors = [this.root]
    for (const char of word) {
      if (!cursors.length) {
        return false
      }
      const nextCursors: TrieNode[] = []
      for (const cursor of cursors) {
        if (char === ".") {
          for (const node of cursor.children) {
            if (node) {
              nextCursors.push(node)
            }
          }
        } else {
          const index = CHAR_INDEX[char]
          if (cursor.children[index]) {
            nextCursors.push(cursor.children[index]!)
          }
        }
      }
      cursors = nextCursors
    }
    return !!cursors.find((c) => c.isTerminator)
  }
}

File: 211_design_add_search_words_data_structure.py
from __future__ import annotations


class TrieNode:
    children: list[TrieNode | None]
    isEnd: bool

    def __init__(self):
        self.children = [None] * 26
        self.isEnd = False


class WordDictionary:
    root: TrieNode
    char_indexes = {char: i for i, char in enumerate("abcdefghijklmnopqrstuvwxyz")}

    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        node = self.root
        for char in word:
            index = self.char_indexes[char]
            if node.children[index] is None:
                node.children[index] = TrieNode()
            node = node.children[index]
        node.isEnd = True

    def search(self, word: str) -> bool:
        cursors = [self.root]
        for char in word:
            next_cursors = []

            for cursor in cursors:
                if char == ".":
                    for node in cursor.children:
                        if node:
                            next_cursors.append(node)
                else:
                    index = self.char_indexes[char]
                    if cursor.children[index]:
                        next_cursors.append(cursor.children[index])

            if not next_cursors:
                return False

            cursors = next_cursors

        matching_cursor = next((c for c in cursors if c.isEnd), None)
        if matching_cursor:
            return True
        else:
            return False
